src/input/yaml.rs

```
@@ -0,0 +1,159 @@
+//! YAML input processing functionality for Sprocket.
+
+use std::fs;
+use std::path::PathBuf;
+
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::bail;
+use tracing::debug;
+
+/// Identifies the format of an input file.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum InputFormat {
+    /// JSON format
+    Json,
+    /// YAML format
+    Yaml,
+}
+
+/// Parses an input file as either JSON or YAML.
+///
+/// Tries to parse as JSON first, then YAML if JSON parsing fails.
+/// Returns the parsed data as a JSON value along with the detected format.
+pub fn parse_input_file(path: &PathBuf) -> Result<(InputFormat, serde_json::Value)> {
```

While this approach does work, it makes it harder to discover errors in the input file and requires several different parses of the input file when the input file is YAML.

An alternative to this approach would be:

* Determine the input file type by extension, supporting `json`, `yaml`, and `yml`, as those are the most popular extensions for these file types.
* Call `Inputs::parse_json` or `Inputs::parse_yaml` depending on the file extension.
* In `wdl-engine`, rename `Inputs::parse` to `Inputs::parse_json`.
* In `wdl-engine`, implement a new method `Inputs::parse_yaml` that uses `serde-yaml-ng` to parse the file into a `serde_yaml_ng::Value` and calls `Inputs::parse_object`, much like `parse_json` does.
* Change `Inputs::parse_object` (and other methods it calls) to take a generic type parameter with a `: Map` bounds (see trait definition below) for abstracting the common operations needed.

It can be tricky to abstract the common functionality to a trait, but looking at `wdl-engine/src/inputs.rs`, generally we need the following:

```rust
trait Map {
    type Value: TryIntoValue;

    // These methods return `Result` items as YAML map keys aren't always string, unlike JSON
    // In the YAML implementation, if a key isn't a string, we return an error.
    fn iter(&self) -> impl Iterator<Item = Result<(&str, &Self::Value)>>;
    fn into_iter(self) -> impl Iterator<Item = Result<(String, Self::Value)>>;
}

impl Map for serde_json::Map<String, serde_json::Value> {
    type Value = serde_json::Value;

    // Implementation here
}

impl Map for serde_yaml_ng::Mapping {
    type Value = serde_yaml_ng::Value;

    // Implementation here
}

trait TryIntoValue {
    fn try_into(self) -> Result<wdl_engine::Value>;
}

impl TryIntoValue for serde_json::Value {
    // Implementation here using `serde_json::from_value`.
}

impl TryIntoValue for serde_yaml_ng::Value {
    // Implementation here using `serde_yaml_ng::from_value`
}
```

---

@peterhuene I thought that we didn't want to add YAML support to the `wdl` crates? I advised this approach to @Gyan-max with that in mind.

---

I am the one to blame for that as I believe I advised you that this would be better implemented in Sprocket rather than in `wdl-engine` based solely on JSON being the only "official" input format in the WDL specification.

However, seeing the burden that decision places on the implementation in Sprocket (especially having to parse files multiple times and secretly converting YAML to JSON on disk to feed it into `wdl-engine`, coupled with the potential to swallow more meaningful errors in the input files), I'm left with the belief that we should elevate YAML to first class input support in `wdl-engine`.

I apologize if this caused undue work as I should have thought ahead as to what it meant for the current `wdl-engine` API to remain unchanged.

---

Peter, can you check out https://github.com/stjude-rust-labs/wdl/issues/347 ?

I made some suggestions there based on the premise we didn't want to add YAML support.